### PR TITLE
fixes for command VS Code API tests

### LINKED ISSUE #4230: complete implementation of commands APIs
- [x] `getCommands`
- [x] `registerTextEditorCommand`
- [x] `thisArg` missing for command registration

### LINKED ISSUE #4241: document.openDocument cannot open URIs properly

### LINKED ISSUE #4242: documents.showTextDocuments resolves before the document is actually focused
It leads to race conditions in VS Code extension code.

### PR DESCRIPTION
**Pull request by akosyakov**
fix #4230, fix #4241, fix #4242

### CODE REVIEW

File: packages/plugin-ext/src/plugin/command-registry.ts [hunk 1]
```diff
@@ -57,7 +57,8 @@ export class CommandRegistryImpl implements CommandRegistryExt {
         }
     }
 
-    registerCommand(command: theia.Command, handler?: Handler): Disposable {
+    // tslint:disable-next-line:no-any
+    registerCommand(command: theia.Command, handler?: Handler, thisArg?: any): Disposable {
```
[the remark this answers is not in the document]
Comment (benoitf) [reply packages/plugin-ext/src/plugin/command-registry.ts]: I mean, in this implementation.
I agree on vscode.d.ts but I don't see how we're using it if it's given in this method

File: packages/plugin-ext/src/plugin/command-registry.ts [hunk 2]
```diff
@@ -134,6 +137,15 @@ export class CommandRegistryImpl implements CommandRegistryExt {
         }
         return commands.executeCommand(actualCmd.command ? actualCmd.command : actualCmd.id, ...(actualCmd.arguments || []));
     }
+
+    getCommands(filterUnderscoreCommands: boolean = false): Thenable<string[]> {
+        return this.proxy.$getCommands().then(result => {
```
[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/plugin-ext/src/plugin/command-registry.ts]: ok

File: packages/plugin-ext/src/plugin/command-registry.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/plugin-ext/src/plugin/command-registry.ts]: fixed

**Comment by akosyakov** [comment]
@benoitf I've addressed what i could.

- i don't think it is a good idea to use `external` property of `vscode-uri`, it is not API. In tests documents are opened with a path, not a URI, and then derived URIs don't have this property at all. `URI.revive` seems to handle all cases.
- logging on `registerTextEditorCommand` is aligned with VS Code implementations

**Pull request merged**